### DISCUSSION TITLE
mpi4py compatibility with `subprocess.run(["mpirun"]...)`

### DISCUSSION THREAD

**Discussion by cwschilly** [mpi4py/mpi4py#533] [Q&A]
## Set Up

I have the following Python script:

```python
import subprocess
subprocess.run(["mpirun", "-n", "4", "./mpi_main"],
               check=True,
               text=True,
               capture_output=True)
```

where `mpi_main` is a very simple MPI executable that runs properly if I call `mpirun -n 4 ./mpi_main` from the command line.

## Problem

This Python script works fine (ie it executes `mpi_main` as expected) when `mpi4py` is NOT installed in my virtual environment.

If `mpi4py` IS installed in the virtual environment, the script fails with the following error message:

```
subprocess.CalledProcessError: Command ["mpirun", "-n", "4", "./mpi_main"] returned non-zero exit status 1
```

## Environment

I am using:

- python=3.9.16
- mpi4py=4.0.0
- OpenMPI=4.0.5
- gcc=10.2.0

I installed mpi4py with the following command:

```
MPI4PY_BUILD_MPICC=/path/to/correct/mpicc pip install --no-cache-dir mpi4py
```
---
@fnrizzi fyi

**Comment by dalcinl** [comment]
Are you running on Linux? mpi4py should not affect AT ALL running an MPI binary executable with mpiexec/mpirun. I have no clue what's going on, but I would really need exacts steps to reproduce the problem, even better if these steps are performed in a Docker container starting from a Docker image published in a public registry.

**Reply by cwschilly** [reply]
Unfortunately, I am unable to provide a Docker container. Generally speaking, could this stem from an issue with firewalls/proxies?

**Reply by dalcinl** [reply]
> I am unable to provide a Docker container.

I was not asking you to provide a Docker container, but rather the sequence of commands that you would use inside a container to setup your Python environment. I'm just trying to reproduce the issue.

> could this stem from an issue with firewalls/proxies?

Very unlinkely. Just by installing a package, things should work. The only reason this would fail is somehow  mpi4py was already imported (indirectly) by the time you execute `subprocess.run(...)`. Or perhaps your `mpi_main` executable calls Python code that uses mpi4py?

Could you please add the following lines to the BEGINNING of your script>
```python
import sys; print(sys.environ.get("mpi4py"), flush=True) # should print `None`
```

Also, what's in your `mpi_main` executable? Does it still file if you  use a basic [helloworld program implemented in C](https://github.com/mpi4py/mpi4py/blob/master/demo/helloworld.c)?

**Reply by cwschilly** [reply]
Thanks for the response @dalcinl

1. Here are the steps to create my Python environment:

```sh
python -m venv mpi4py_test
source mpi4py_test/bin/activate
pip install mpi4py
```

2. Running the code you sent (switching out `sys` for `os`) indeed prints out `None`.

3. My `mpi_main` just distributes a vector and does some random operations on it (no Python code involved). I tried with the `helloworld.c` you sent, and it fails with the same error:

```
subprocess.CalledProcessError: Command ["mpirun", "-n", "4", "./hello_world"] returned non-zero exit status 1
```

**Reply by dalcinl** [reply]
What OS and version are you using? What about MPI? What MPI are you using? Are you using a system-provided MPI (e.g installed with dnf or apt) or a custom build? If the second, how did you build your MPI? Without more information, it is impossible to even guess what's going on.

**Reply by cwschilly** [reply]
I am running on RHEL 8, with a system-provided MPI (OpenMPI 4.1.6).

**Comment by dalcinl** [comment]
I tried to reproduce in a RockyLinux container:
```console
$ podman run -it --rm rockylinux:8
```
Then I executed the following:
```bash
dnf install -y python39-devel openmpi-devel
source /etc/profile.d/modules.sh
module load mpi/openmpi-x86_64
export OMPI_ALLOW_RUN_AS_ROOT=1
export OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1

cat <<EOF > test.c
#include <stdio.h>
#include <mpi.h>
int main(int argc, char *argv[])
{
  int size, rank, len;
  char name[MPI_MAX_PROCESSOR_NAME];
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Get_processor_name(name, &len);
  printf("Hello, World! I am process %d of %d on %s.\n", rank, size, name);
  MPI_Finalize();
  return 0;
}
EOF

cat <<EOF > test.py
import subprocess
subprocess.run(["mpirun", "-n", "4", "./a.out"],
               check=True,
               text=True,
               capture_output=True)
EOF

mpicc test.c
python3 test.py # first execution, no mpi4py

python3 -m venv test
source test/bin/activate
pip install mpi4py

python3 test.py # second execution, mpi4py installed in venv
```
and I get no errors the two times `python test.py` runs. 
Note however the Python and MPI versions are a bit different.

**Reply by cwschilly** [reply]
Hi @dalcinl 

Thanks again for helping me try to debug this. I'm not sure what the root problem was/is, but using MPICH (built from source) instead of OpenMPI fixed things for me.

**Comment by heplesser** [comment]
This seems to be an issue with `subprocess`, not with mpi4py, see https://github.com/python/cpython/issues/132377.

**Reply by dalcinl** [reply]
IIRC, Open MPI errors is you try to exec `mpirun` from a process that already called `MPI_Init`, and that's exactly what your code in python/cpython#132377 seems to be doing. I believe the problem is not in Python, but in a constraint from the MPI runtime.
Do you get any other output than the exception traceback? Any messages about the code using `fork()` ?

**Reply by heplesser** [reply]
I haven't been able to get any detailed diagnostic output. But I found now that the problem no longer occurs with OpenMPI 5.0.7, neither under Linux nor macOS, so I assume it was an OpenMPI problem. The OpenMPI change log for 5.0.4 mentions "Fix a bug in intercomm creation", maybe that was the problem. This would also chime with the report by @cwschilly that MPICH worked.
